src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/FileServerHandler.java

```diff
@@ -265,25 +269,156 @@ private void serveDefaultFavIcon(HttpExchange exchange, boolean writeBody)
             exchange.sendResponseHeaders(200, -1);
+            return;
+        }
+        String rangeHeader = exchange.getRequestHeaders().getFirst("Range");
+        if (rangeHeader != null && handleRangedRequest(exchange, path, rangeHeader)) {
+            return;
+        }
+        exchange.sendResponseHeaders(200, Files.size(path));
+        try (InputStream fis = Files.newInputStream(path);
+             OutputStream os = exchange.getResponseBody()) {
+            fis.transferTo(os);
+        }
+    }
+
+    private boolean handleRangedRequest(HttpExchange exchange, Path path, String rangeHeader)
+            throws IOException
+    {
+        var reqHdrs = exchange.getRequestHeaders();
+        String ifRange = reqHdrs.getFirst("If-Range");
+        if (!(ifRange == null || ifRange.equals(getLastModified(path)))) {
+            return false; // last-modified does not match with the resource, send the entire file
+        }
+        long fileSize = Files.size(path);
+        List<RangeEntry> ranges = parseRangeHeader(rangeHeader, fileSize);
+        if (ranges == null) {
+            var respHdrs = exchange.getResponseHeaders();
+            respHdrs.set("Content-Range", "bytes */%s".formatted(fileSize));
+            exchange.sendResponseHeaders(416, -1);
+            return true;
+        }
+        servePartialContents(exchange, path, ranges);
+        return true;
+    }
+
+    // Parses the HTTP Range header and returns a list of byte ranges.
+    // Returns null if the header is invalid.
+    // Example of valid Range header values:make image
+    //  "bytes=0-499"           -> first 500 bytes
+    //  "bytes=500-999"         -> second 500 bytes
+    //  "bytes=-500"            -> last 500 bytes
+    //  "bytes=9500-"           -> from byte 9500 to end
+    // Multiple ranges are allowed, separated by commas(without spaces):
+    //  "bytes=0-499,1000-1499" -> first 500 bytes and second 500 bytes
+    public static List<RangeEntry> parseRangeHeader(String rangeHeader, long fileSize) {
+        if (!rangeHeader.startsWith("bytes=")) {
+            return null;  // only 'bytes' unit is supported.
+        }
+        String rangesPart = rangeHeader.substring("bytes=".length());
+        List<RangeEntry> ranges = new ArrayList<>();
+        for (String spec : rangesPart.split(",", -1)) {
+            spec = spec.trim();
+            int dashPos = spec.indexOf('-');
+            if (dashPos == -1) {
+                return null;  // '-' is always required.
+            }
+            String startStr = spec.substring(0, dashPos).trim();
+            String endStr = spec.substring(dashPos + 1).trim();
+            if (startStr.startsWith("+") || endStr.startsWith("+")) {
+                return null;  // "+" prefix is not supported.
+            }
+            long start, end;
+            try {
+                if (startStr.isEmpty()) {  // "-<suffix-length>"
+                    long suffixLength = Long.parseLong(endStr, 10);
```

Comment:
`Long.parseLong(endStr)` already parses in base 10, there's no need to be explicit about it. See the JDK code:

    public static long parseLong(String s) throws NumberFormatException {
        return parseLong(s, 10);
    }